### ISSUE TITLE
Run logit experimentations for conversational collapse

### ISSUE THREAD

**Issue by cemrtkn** [cemrtkn/conversational-collapse#1]
# Overview

Conversational collapse happens when one or more language models talk to each other for a large amount of rounds. It shows in the output of these models which becomes increasingly similar to the previous message in the context. This holds for most of the setups with *some exceptions* to the extent that models start repeating the previous message word by word or paraphrase them.

*exceptions can be studied later*

In this context, the first experiment that would reveal information about what is going on inside a model when this collapse happens is looking at logits. Models seems to be "sure" about exactly what words follow as the conversation keeps going and it would be nice to show that this happens. 

The expectation is that the model will become more and more overconfident and this will display itself in the logit distribution losing entropy.
## Specifications
- [x] Change the interpretable inference module so that it allows modules of NNSight code that injects into the generation setup. This should not affect the output generation. The injection code should look like this:
```python
from api.interp_inference import DEFAULT_MODEL, InterpInference


def save_last_layer_activations(model):
    """Intervention that saves the last layer's hidden states."""
    # Access the last transformer layer and save its output
    last_layer_output = model.model.layers[-1].output[0].save()
    return {"last_layer": last_layer_output}


def main():
    inference = InterpInference(model=DEFAULT_MODEL)

    response, interp_outputs = inference.generate_from_messages(
        messages=[
            {
                "role": "user",
                "content": "What is the capital of France?",
            }
        ],
        max_new_tokens=100,
        intervention=save_last_layer_activations,
    )
    
    print(f"Response: {response.content}")
    
    if interp_outputs:
        last_layer = interp_outputs["last_layer"]
        print(f"Last layer shape: {last_layer.shape}")


if __name__ == "__main__":
    main()
````
- [x] Run experiment by sampling logits in every step of the generation and save results.
- [x] Visualize results along a time axis

**Comment by cemrtkn** [comment]
# Overview
After implementing an nnsight-based inference that allows for interventions, I conducted collapse experiments with various conversation samples as starters. The overconfidence per conversation round was proxied through avg. entropy of each generation step in that round. Low entropy means more over-confidence. Only 3 different conversation starters were used. Results were visualized through two graphs:

1. y = avg_entropy & x=conversation_round
2. First 80 characters of each round's results in text listed vertically. Color-coded based on avg_entropy for that round. 


## Observations
- The results showed that for most cases, there was considerable overconfidence in the logits. However, the confidence was not consistently rising as the conversation kept going. 
- There were cases where, even though a text that is very similar to the previous round was generated, the avg_entropy increased.
- The entropy trend varied from one topic to another.
